x/bsonx/bsoncore/document_test.go

```diff
@@ -410,3 +412,128 @@ func TestDocument(t *testing.T) {
 		}
 	})
 }
+
+func TestDocument_StringN(t *testing.T) {
+	var buf strings.Builder
+	for i := 0; i < 16000000; i++ {
+		buf.WriteString("abcdefgh")
+	}
+	str1k := buf.String()
+	str128 := str1k[:128]
+
+	testCases := []struct {
+		description string
+		n           int
+		doc         Document
+		want        string
+	}{
+		// n = 0 cases
+		{"n=0, document with 1 field", 0, BuildDocument(nil,
+			AppendStringElement(nil, "key", str128),
+		), ""},
+
+		{"n=0, empty document", 0, Document{}, ""},
+
+		{"n=0, document with nested documents", 0, BuildDocument(nil,
+			AppendDocumentElement(nil, "key", BuildDocument(nil,
+				AppendStringElement(nil, "nestedKey", str128),
+			)),
+		), ""},
+
+		{"n=0, document with mixed types", 0, BuildDocument(nil,
+			AppendStringElement(nil, "key", str128),
+			AppendInt32Element(nil, "number", 123),
+		), ""},
+
+		{"n=0, deeply nested document", 0, BuildDocument(nil,
+			AppendDocumentElement(nil, "a", BuildDocument(nil,
+				AppendDocumentElement(nil, "b", BuildDocument(nil,
+					AppendStringElement(nil, "c", str128),
+				)),
+			)),
+		), ""},
+
+		{"n=0, complex value", 0, BuildDocument(nil,
+			AppendDocumentElement(nil, "key", BuildDocument(nil,
+				AppendStringElement(nil, "nestedKey", str128),
+			)),
+		), ""},
+
+		// n < 0 cases
+		{"n<0, document with 1 field", -1, BuildDocument(nil,
+			AppendStringElement(nil, "key", str128),
+		), ""},
+
+		{"n<0, empty document", -1, Document{}, ""},
+
+		{"n<0, document with nested documents", -1, BuildDocument(nil,
+			AppendDocumentElement(nil, "key", BuildDocument(nil,
+				AppendStringElement(nil, "nestedKey", str128),
+			)),
+		), ""},
+
+		{"n<0, document with mixed types", -1, BuildDocument(nil,
+			AppendStringElement(nil, "key", str128),
+			AppendInt32Element(nil, "number", 123),
+		), ""},
+
+		{"n<0, deeply nested document", -1, BuildDocument(nil,
+			AppendDocumentElement(nil, "a", BuildDocument(nil,
+				AppendDocumentElement(nil, "b", BuildDocument(nil,
+					AppendStringElement(nil, "c", str128),
+				)),
+			)),
+		), ""},
+
+		{"n<0, complex value", -1, BuildDocument(nil,
+			AppendDocumentElement(nil, "key", BuildDocument(nil,
+				AppendStringElement(nil, "nestedKey", str128),
+			)),
+		), ""},
+
+		// n > 0 cases
+		{"n>0, document LT n", 3, BuildDocument(nil,
+			AppendStringElement(nil, "key", "value"),
+		), `{"}`},
```

Comment:
I would not expect this to be `{"}`, I would expect `{"\k`. To resolve this, make the truncated update suggested the other commend and remove this line from the algorithm:

```
n -= buf.Len()
```

There is no need to change the value of `n` before determining truncation. buf.Len() will always be 1 byte at this stage (`{`}.